### PR TITLE
support sm2 for ioctl's account cmds

### PR DESCRIPTION
**Pull request by Frankonly**
Support sm2 in ioctl and create `xctl.go` in `tools\ioctl\xctl` who using sm2 as default.

### CODE REVIEW

File: ioctl/cmd/account/account.go [hunk 1]
```diff
@@ -57,6 +59,13 @@ var (
 	ErrPasswdNotMatch = errors.New("password doesn't match")
 )
 
+var cryptoSm2 bool
+
+var flagSm2Usage = map[config.Language]string{
+	config.English: "sm2 cryptographic algorithm",
+	config.Chinese: "国密sm2加密算法",
```
Comment (dustinxie) [inline ioctl/cmd/account/account.go]: 签名算法

File: ioctl/cmd/account/account.go [hunk 2]
```diff
@@ -131,6 +137,19 @@ func KsAccountToPrivateKey(signer, password string) (crypto.PrivateKey, error) {
...
+
+	// find the account in pem files
+	pemFilePath := filepath.Join(config.ReadConfig.Wallet, "sm2sk-"+address.String()+".pem")
+	prvKey, err := crypto.ReadPrivateKeyFromPem(pemFilePath, password)
+	if err == nil {
+		return prvKey, nil
+	}
+
+	var pathError *os.PathError
+	if !errors.As(err, &pathError) {
```
Comment (dustinxie) [inline ioctl/cmd/account/account.go]: we don't need this error?

it failed to find in keystore, and no sm2 exist, so it's just "not match all keys in keystore", not failed to read sm2 key

File: ioctl/cmd/account/account.go (hunk 1 above)
Comment (dustinxie) [reply ioctl/cmd/account/account.go]: also, can you resolve the 2 //TODO: in accountimport.go? delete they if not needed

File: ioctl/cmd/account/account.go [hunk 3]
```diff
@@ -228,30 +274,41 @@ func newAccountByKeyStore(alias, passwordOfKeyStore, keyStorePath string, wallet
...
+}
+
 func storeKey(privateKey, walletDir, password string) (string, error) {
 	priKey, err := crypto.HexStringToPrivateKey(privateKey)
 	if err != nil {
 		return "", output.NewError(output.CryptoError, "failed to generate private key from hex string ", err)
 	}
 	defer priKey.Zero()
 
+	addr, err := address.FromBytes(priKey.PublicKey().Hash())
```
Comment (dustinxie) [inline ioctl/cmd/account/account.go]: not necessary?
HexStringToPrivateKey() will return error if the string cannot make a valid private key

Comment (dustinxie) [reply ioctl/cmd/account/account.go]: i see, move from below

File: ioctl/cmd/account/account_test.go [hunk 4]
```diff
@@ -37,21 +38,37 @@ func TestAccount(t *testing.T) {
...
+	prvKey2, err := LocalAccountToPrivateKey(addr2.String(), passwd)
+	require.NoError(err)
+	msg2 := hash.Hash256b([]byte(nonce))
+	sig2, err := prvKey2.Sign(msg2[:])
 	require.NoError(err)
-	require.True(prvkey.PublicKey().Verify(msg[:], sig))
+	require.True(prvKey2.PublicKey().Verify(msg2[:], sig2))
 
 	// test import existing key
 	sk, err := crypto.GenerateKey()
```
Comment (dustinxie) [inline ioctl/cmd/account/account_test.go]: add test to load from PEM file

File: ioctl/cmd/account/accountlist.go [hunk 5]
```diff
@@ -8,15 +8,18 @@ package account
 
 import (
 	"fmt"
+	"io/ioutil"
 	"strings"
 
 	"github.com/ethereum/go-ethereum/accounts/keystore"
 	"github.com/spf13/cobra"
```
Comment (dustinxie) [inline ioctl/cmd/account/accountlist.go]: delete line

File: ioctl/cmd/account/accountlist.go [hunk 6]
```diff
@@ -67,6 +71,23 @@ func accountList() error {
 			Alias:   aliases[address.String()],
 		})
 	}
+	files, err := ioutil.ReadDir(config.ReadConfig.Wallet)
```
Comment (dustinxie) [inline ioctl/cmd/account/accountlist.go]: can we add test for these new code?

File: ioctl/cmd/account/accountupdate.go [hunk 7]
```diff
@@ -62,33 +66,72 @@ func accountUpdate(arg string) error {
...
 				return output.NewError(output.KeystoreError, "failed to update keystore", err)
 			}
+
 			output.PrintResult(fmt.Sprintf("Account #%s has been updated.", account))
+			return nil
 		}
 	}
-	return output.NewError(output.KeystoreError, fmt.Sprintf("account #%s not found", account), nil)
+
+	// find the pem file and update
```
Comment (dustinxie) [inline ioctl/cmd/account/accountupdate.go]: can we add test for these new code?

File: tools/ioctl/xctl/xctl.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+// License 2.0 that can be found in the LICENSE file.
+
+package main
+
+import (
+	"github.com/iotexproject/iotex-core/ioctl/cmd"
+	"github.com/iotexproject/iotex-core/ioctl/cmd/account"
+)
+
+func main() {
```
Comment (dustinxie) [inline tools/ioctl/xctl/xctl.go]: add an entry in makefile to generate binary with diff name?

File: ioctl/cmd/account/account_test.go (hunk 4 above)
Comment (Frankonly) [reply ioctl/cmd/account/account_test.go]: The case that loading SM2 key from PEM file is included when running `LocalAccountToPrivateKey()`.

File: ioctl/cmd/account/accountlist.go (hunk 5 above)
Comment (Frankonly) [reply ioctl/cmd/account/accountlist.go]: `io/ioutil` is used at line 74:
```
files, err := ioutil.ReadDir(config.ReadConfig.Wallet)
```

File: ioctl/cmd/account/accountcreate.go [hunk 9]
```diff
@@ -61,15 +61,28 @@ type generatedAccount struct {
 }
 
 func init() {
-	accountCreateCmd.Flags().UintVarP(&numAccounts, "num", "n", 1, config.TranslateInLang(flagNumUsages, config.UILanguage))
+	accountCreateCmd.Flags().UintVarP(&numAccounts, "num", "n", 1,
+		config.TranslateInLang(flagNumUsages, config.UILanguage))
+	accountCreateCmd.Flags().BoolVar(&CryptoSm2, "sm2", false,
+		config.TranslateInLang(flagSm2Usage, config.UILanguage))
```
Comment (CoderZhi) [inline ioctl/cmd/account/accountcreate.go]: why we need a flag?

File: ioctl/cmd/account/accountlist.go [hunk 10]
```diff
@@ -44,7 +45,8 @@ var accountListCmd = &cobra.Command{
 }
 
 type listMessage struct {
-	Accounts []account `json:"accounts"`
+	Accounts    []account `json:"accounts"`
+	Sm2Accounts []account `json:"sm2Accounts"`
```
Comment (CoderZhi) [inline ioctl/cmd/account/accountlist.go]: why we need to distinguish?

File: ioctl/cmd/account/accountcreate.go (hunk 9 above)
Comment (Frankonly) [reply ioctl/cmd/account/accountcreate.go]: > why we need a flag?

I make ioctl support both two cryptography algorithms. This flag is to swtich them when using command like `account create`, avoiding more commands for similar functions.

File: ioctl/cmd/account/accountlist.go (hunk 10 above)
Comment (Frankonly) [reply ioctl/cmd/account/accountlist.go]: > why we need to distinguish?

ioctl can handle different keys with different cryptography algorithms, if user stores those two kinds of key in same place (using same ioctl), ioctl need distinguish them when listing them for user.

File: ioctl/cmd/account/accountcreate.go (hunk 9 above)
Comment (Frankonly) [reply ioctl/cmd/account/accountcreate.go]: ioctl supports both two cryptography algorithms. xctl only supports sm2. How I implement xctl is to set `CryptoSm2` to `true` always.

**Pull request merged**